### PR TITLE
Coerce MxNx1 images into MxN images for imshow

### LINKED ISSUE #15089: Coerce MxNx1 images into MxN images for imshow
Not a bug report and possibly a repeat or bad idea, but using pytorch, keras, theano, tensorflow you often end up with images with shape (x, y, 1). It would be nice, if matplotlib imshow would just assume images of that size are grayscale rather than showing an error message.

```
import matplotlib.pyplot as plt
import numpy as np
gray_image = np.ones((3, 3, 1))
plt.imshow(gray_image)
```

# Result

```
TypeError: Invalid shape (3, 3, 1) for image data
```
https://github.com/matplotlib/matplotlib/blob/0290a7e2382598cf3c54150abcf4e55a0b455941/lib/matplotlib/image.py#L674-L677

# Expected Result

The same as if I did
```
plt.imshow(gray_image[:, :, 0])
````

### PR DESCRIPTION
**Pull request by kmader**
## PR Summary
Causes `matplotlib.pyplot.imshow` to work with depth=1 images (common in deep learning and segmentation) as if they were grayscale rather than throwing an error message (so it should be backward compatible, unless someone expected it to throw an error)
- closes #15089
## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

## Specific Tests

- [x] shape (n,m) -> colormapping
- [x] shape(n,m,1) -> colormapping, same as (n,m), newly introduced here.
- [x] shape(n,m,2) -> Error
- [x] shape(n,m,3) -> RGB image
- [x] shape(n,m,4) -> RGBA image
- [x] shape(n,m,5) -> Error

### CODE REVIEW

**Comment by kmader** [comment]
I could easily add a unit test to `test_image.py`, but wasn't sure if it was better to add a new one or piggy back it on another test that already does grayscale images.

File: lib/matplotlib/image.py [hunk 1]
```diff
@@ -671,6 +671,10 @@ def set_data(self, A):
             raise TypeError("Image data of dtype {} cannot be converted to "
                             "float".format(self._A.dtype))
 
+        if not (self._A.ndim == 3 and self._A.shape[-1]==1):
```
Comment (tacaswell) [inline lib/matplotlib/image.py]:
```suggestion
        if (self._A.ndim == 3 and self._A.shape[-1]==1):
```

I think this logic is inverted?

**Comment by tacaswell** [comment]
This is projecting (N, M, 1) input down to (N, M) which means in goes through color mapping code is by default produces an image which is not grayscale.  The other option would be to project it _up_ to a (N, M, 3) array where each plane is identical which would run the input through the RGB(A) handling code path which would produce a gray-scale output.

My inclination would be towards the first case, but from your language motivating this change I expected it to be doing the second.

**Comment by tacaswell** [comment]
Thank you for working on this!

Either way, we will definitely need a test and a whats_new or API change entry.

File: lib/matplotlib/image.py (hunk 1 above)
Comment (kmader) [reply lib/matplotlib/image.py]: sorry! you are right!

**Comment by kmader** [comment]
@tacaswell @jklymak for the usual use cases with tensorflow and pytorch it would make more sense to treat it the same way as the N, M (rather than N, M, 3) since it is often scaled differently. For me the main point would be that all model outputs could be handled the same way rather than having add `if`s around every statement
`if x.ndim==3 and x.shape[-1]==1 elif x.ndim==3 and x.shape[-1]==3`

**Comment by tacaswell** [comment]
For your current code, `ax.imshow(img.squeeze(),...)` should work in most cases (it would behave badly if you have shape (1, N) or (M, 1) input so we can't do that internally, but I suspect that in 99% of image based machine learning workflows that would be ok).

**Comment by kmader** [comment]
@tacaswell that is potentially a perfectly acceptable solution.Should I close the PR?
If we close it,  I assume (maybe wrongly) that other people might run into this issue, would it make sense to give them more directed feedback on how to use imshow correctly when the input is MxNx1?

**Comment by jklymak** [comment]
We reshape arrays in other contexts.  I think coercing MxNx1 to MxN is reasonable in this case.

File: doc/users/next_whats_new/imshow_m_n_1.rst [hunk 2]
```diff
@@ -0,0 +1,6 @@
+Imshow now coerces 3D arrays with depth 1 to 2D
+------------------------------------------------
+The number of ticks on colorbars was appropriate for a large colorbar, but
```
Comment (ImportanceOfBeingErnest) [inline doc/users/next_whats_new/imshow_m_n_1.rst]: The number of ticks on colorbars seems unrelated.

File: doc/users/next_whats_new/imshow_m_n_1.rst [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+Starting from this version arrays of size MxNx1 will be coerced into MxN 
+for displaying. This means commands like ``plt.imshow(np.ones((3, 3, 1)))``
```
Comment (ImportanceOfBeingErnest) [inline doc/users/next_whats_new/imshow_m_n_1.rst]: `np.ones((3, 3, 1))` may be bad example because colormapping will give unexpected results for identical values. Maybe `np.random.rand(3, 3, 1)` instead?

File: doc/users/next_whats_new/imshow_m_n_1.rst (hunk 2 above)
Comment (kmader) [reply doc/users/next_whats_new/imshow_m_n_1.rst]: sorry! I hadn't seen that from the copy paste

**Comment by ImportanceOfBeingErnest** [comment]
I think ultimately one would like to test each of the following cases

1. shape (n,m) -> colormapping
2. shape(n,m,1) -> colormapping, same as (n,m), newly introduced here.
3. shape(n,m,2) -> Error
4. shape(n,m,3) -> RGB image
5. shape(n,m,4) -> RGBA image
6. shape(n,m,5) -> Error

It doesn't look like those are all tested currently. 

An image comparisson test may make sense for cases 1, 4 and 5. (All of those could be tested within the same figure) For case 2. one may rather compare the output to case 1., using the [`check_figures_equal`](https://matplotlib.org/api/testing_api.html#matplotlib.testing.decorators.check_figures_equal) decorator. (We do in general try to keep the number of image comparison tests as low as possible.)

**Comment by kmader** [comment]
@ImportanceOfBeingErnest I hadn't known about the `check_figures_equal` decorator that probably makes more sense for the cases. I already added the check that imshow gives the same result with MxNx1 and MxN (case 2). Should I add checks for all the other cases as well?

**Comment by ImportanceOfBeingErnest** [comment]
That would be the ultimate goal. Up to you if you want to do that work and include it in this PR or not.

**Comment by kmader** [comment]
The x3 and x4 cases seem to be already covered in other tests, so that should 
https://github.com/matplotlib/matplotlib/blob/22a339242af761527e786da243361d9a579127ac/lib/matplotlib/tests/test_image.py#L49-L59

https://github.com/matplotlib/matplotlib/blob/22a339242af761527e786da243361d9a579127ac/lib/matplotlib/tests/test_image.py#L276-L298

**Comment by kmader** [comment]
@ImportanceOfBeingErnest everything seems fine for me (but I am very inclined to say that), anything else I should add?

**Comment by ImportanceOfBeingErnest** [comment]
Appart from the flake8 error this looks fine to me.  
A general question would be if we want to have this for `imshow` only. It might be considered as inconsistent to let (n,m,1) arrays be handled by `imshow`, but not by `pcolor[mesh]`.

**Comment by kmader** [comment]
@ImportanceOfBeingErnest I can change it, it appears to be a completely different mechanism in `matplotlib/axes/_axes.py` but should be possible to change. We should then probably also add tests that explicitly fail for 10x10x2, 10x10x3, and 10x10x4 images because pcolormesh can only ever work on x,y,1 sized images.

**Comment by ImportanceOfBeingErnest** [comment]
I guess that can also be another PR; depending on what is easiest. Approving this as it stands.

File: lib/matplotlib/image.py [hunk 4]
```diff
@@ -671,6 +671,10 @@ def set_data(self, A):
             raise TypeError("Image data of dtype {} cannot be converted to "
                             "float".format(self._A.dtype))
 
+        if (self._A.ndim == 3 and self._A.shape[-1] == 1):
+            # If just one dimension assume grayscale
```
Comment (jklymak) [inline lib/matplotlib/image.py]: This comment isn't correct anymore - its not assuming greyscale, its just doing a mapping into the colormap?

Comment (kmader) [reply lib/matplotlib/image.py]: as that slipped past me, thanks!

**Comment by tacaswell** [comment]
Pre-emptively labeling as release critical to make sure it does not fall through the cracks.

The discussion at https://github.com/matplotlib/matplotlib/issues/7221 suggests that at least some users would expect us to go to RGB gray scale rather than color mapping.

**Comment by jklymak** [comment]
@tacaswell are you suggesting we hold off on merging it?  I'm not sure why MxNx1 would be considered greyscale, but MxN would be considered color mappable.  Is this really a particular convention in machine learning?  I'm not clear on why that would be the case; I can imagine having a 3-D array MxNxNsamps, but I'd have expected that to be accessed as `img[:, :, ind]` which is back to MxN.

**Comment by timhoffm** [comment]
+/-0 on accepting MxNx1. On the one hand, we should have clear APIs. On the other hand, it may be convenient to accept MxNx1 specifically, if people have this kind of data a lot.

However, if MxNx1 is accepted, it should be handled the same way as MxN. I don't think it's safe to assume that MxNx1 is a grayscale image ("In the face of ambiguity, refuse the temptation to guess."). Let's rather go for consistency than a subtle different behavior.

**Comment by kmader** [comment]
My main reason for adding it was my experience from teaching students newer to python where particularly with segmentation problems students were confused at why MxN and MxNx3 images worked but MxNx1 didn't. I can, however, appreciate that there are few easy decisions.

**Comment by jklymak** [comment]
Right but how did they get an mxnx1 matrix in the first place?

**Comment by kmader** [comment]
The main reason (AFAIK) they do this is to make convolutions in convolutional neural networks easier. So all the convolutions have MxNxQ to MxNxR format and binary is just when R=1

**Comment by NelleV** [comment]
Looks good to me as well. I also asked @jni as someone who works on images if he thought there was a case where we wouldn't want this, and he couldn't find any.
